src/Compilers/CSharp/Portable/Parser/Lexer.cs

```diff
@@ -925,11 +925,43 @@ private void ScanSyntaxToken(ref TokenInfo info)
                         info.Text = TextWindow.GetText(intern: true);
                     }
 
-                    this.AddError(ErrorCode.ERR_UnexpectedCharacter, info.Text);
+                    this.AddError(ErrorCode.ERR_UnexpectedCharacter, EscapeIfInvalidUnicode(info.Text));
                     break;
             }
         }
 
+        private static readonly Encoding s_utf8EncodingThatThrows = new UTF8Encoding(encoderShouldEmitUTF8Identifier: false, throwOnInvalidBytes: true);
+
+        private static string EscapeIfInvalidUnicode(string text)
+        {
+            // This is an unfortunate hack, but is reasonably simple and unblocks an unfortunate situation.
+            // Specifically, there are consumers of Roslyn data that end up needing to encode the strings we produce
+            // (for example in diagnostic messages) from legal .Net strings to things like utf8.  For example, the
+            // rpc-serialization libraries in play have an additional restriction that .Net string be valid utf8, or
+            // else they will lose data (non unicode characters get replaced with \uFFFD).
+            //
+            // To mitigate this issue, if we're about to report an issue about a surrogate that we ourselves is invalid
+            // for C#, then we don't report the error about that char directly (despite it being legal .Net), and instead
+            // encode the character into a form that later encoders/decoders won't have an issue with.
+            try
+            {
+                s_utf8EncodingThatThrows.GetByteCount(text);
```

Comment:
this is the lightest way i could find to do the check.  it involves no allocations, but still fails if we feed in a string that cannot be converted to unicode.

Reply:
You seem to have optimized this well already. Just a suggestion - if you go with char.GetUnicodeCategory over the individual characters in the string, you could still avoid the allocation and optimize the catch (unsure how common that path is hit). At that point, if you really feel scroogy, you could switch to unsafe string and stackalloc and copy uptil the faulty character with a fixed array and append in place.

Reply:
Also `Char.IsSurrogate` and `Char.IsSurrogatePair` can be used.

Reply:
I considered that. But that only helps if the cases that always fail are known to be surrogate cases. It's unclear to me if that's always the case, or if there are other legal chars, which are but surrogates, which we could have a problem with (now or in the future).  So this approach was the most future proof. 

It's legit just asking: can this string be converted?  No? Ok, encode it. :-)